firedrake/plot.py

```diff
@@ -1,373 +1,280 @@
+        interior_kw["linewidths"] = interior_kw.get("linewidths", 1.0)
+        if gdim == 2:
+            interior_kw["facecolors"] = interior_kw.get("facecolors", "none")
+
+        interior_collection = InteriorCollection(vertices, **interior_kw)
+        axes.add_collection(interior_collection)
+        result.append(interior_collection)
+
+    # Add colored lines/polygons for the boundary facets
+    facets = mesh.exterior_facets
+    local_facet_ids = facets.local_facet_dat.data_ro
+    exterior_facet_node_map = coordinates.exterior_facet_node_map().values
+    topology = coordinates.function_space().finat_element.cell.get_topology()
+
+    mask = np.zeros(exterior_facet_node_map.shape, dtype=bool)
+    for facet_index, local_facet_index in enumerate(local_facet_ids):
+        mask[facet_index, topology[tdim - 1][local_facet_index]] = True
+    faces = exterior_facet_node_map[mask].reshape(-1, tdim)
+
+    markers = facets.unique_markers
+    color_key = "colors" if tdim <= 2 else "facecolors"
+    boundary_colors = boundary_kw.pop(color_key, None)
+    if boundary_colors is None:
+        cmap = matplotlib.cm.get_cmap("Dark2")
+        num_markers = len(markers)
+        colors = cmap([k / num_markers for k in range(num_markers)])
+    else:
+        colors = matplotlib.colors.to_rgba_array(boundary_colors)
+
+    boundary_kw = dict(boundary_kw)
+    if tdim == 3:
+        boundary_kw["edgecolors"] = boundary_kw.get("edgecolors", "k")
+        boundary_kw["linewidths"] = boundary_kw.get("linewidths", 1.0)
+    for marker, color in zip(markers, colors):
+        face_indices = facets.subset(int(marker)).indices
+        marker_faces = faces[face_indices, :]
+        vertices = coords[marker_faces]
+        _boundary_kw = dict(**{color_key: color, "label": marker}, **boundary_kw)
+        marker_collection = BoundaryCollection(vertices, **_boundary_kw)
+        axes.add_collection(marker_collection)
+        result.append(marker_collection)
+
+    # Dirty hack to enable legends for 3D volume plots. See the function
+    # `Poly3DCollection.set_3d_properties`.
+    for collection in result:
+        if isinstance(collection, Poly3DCollection):
+            collection._facecolors2d = PolyCollection.get_facecolor(collection)
+            collection._edgecolors2d = PolyCollection.get_edgecolor(collection)
+
+    _autoscale_view(axes, coords)
+    return result
+
+
+def _plot_2d_field(method_name, function, *args, **kwargs):
+    axes = kwargs.pop("axes", None)
     if axes is None:
         figure = plt.figure()
-        axes = figure.add_subplot(111, projection=projection, **kwargs)
+        axes = figure.add_subplot()
 
-    lines = Lines(vertices, colors=colors)
+    if len(function.ufl_shape) == 1:
+        mesh = function.ufl_domain()
+        element = function.ufl_element().sub_elements()[0]
+        Q = FunctionSpace(mesh, element)
+        function = interpolate(sqrt(inner(function, function)), Q)
```

Comment:
That this is what's actually plotted should be noted in the docstring for the functions plot using `_plot_2d_field`